### PR TITLE
API keys: Add migration path for API and terraform to Docs

### PR DESCRIPTION
**Pull request by eleijonmarck**
**What is this feature?**
- adds docs for migration of API and terraform for apikeys and service acconuts

### CODE REVIEW

File: docs/sources/administration/api-keys/index.md [hunk 1]
```diff
@@ -44,28 +48,91 @@ This topic shows you how to create an API key using the Grafana UI. You can also
...
+- The Grafana user interface
+- The Grafana API
+- Terraform
+
+### Migrate API keys to Grafana service accounts using the Grafana user interface
+
+This section shows you how to migrate API keys to Grafana service accounts using the Grafana user interface. You can choose to migrate a single API key or all API keys. When you migrate all API keys, you can no longer create API keys and must use service accounts instead.
 
-- Ensure you have permission to create Grafana service accounts. For more information about permissions, refer to [Roles and permissions]({{< relref "../roles-and-permissions/#" >}}).
+#### Before you begin
```
Comment (chri2547) [inline docs/sources/administration/api-keys/index.md:69]: @eleijonmarck If there are any prerequisites or special permissions a user must have in place, we can add them to this Before you begin section. Are there any such prereqs? If not, we can remove this section.

File: docs/sources/administration/api-keys/index.md [hunk 2]
```diff
@@ -44,28 +48,91 @@ This topic shows you how to create an API key using the Grafana UI. You can also
...
 
 1. Sign in to Grafana, hover your cursor over **Configuration** (the gear icon), and click **API Keys**.
 1. Find the API Key you want to migrate.
 1. Click **Migrate to service account**.
+
+### Migrate API keys to Grafana service accounts using the API
+
+This section shows you how to migrate API keys to Grafana service accounts using the Grafana API.
+
+#### Before you begin
```
Comment (chri2547) [inline docs/sources/administration/api-keys/index.md:95]: @eleijonmarck If there are any prerequisites or special permissions a user must have in place, we can add them to this Before you begin section. Are there any such prereqs? If not, we can remove this section.

File: docs/sources/administration/api-keys/index.md [hunk 3]
```diff
@@ -44,28 +48,91 @@ This topic shows you how to create an API key using the Grafana UI. You can also
...
+
+1. SATs used for authentication
+1. Remove code that handles the old `/api/auth/keys` endpoint.
+1. Track the [API keys](http://localhost:3000/org/apikeys) in use and migrate them to SATs.
+
+### Migrate API keys to Grafana service accounts using Terraform
+
+This section shows you how to migrate API keys to Grafana service accounts using Terraform.
+
+#### Before you begin
```
Comment (chri2547) [inline docs/sources/administration/api-keys/index.md:123]: @eleijonmarck If there are any prerequisites or special permissions a user must have in place, we can add them to this Before you begin section. Are there any such prereqs? If not, we can remove this section.

File: docs/sources/administration/api-keys/index.md [hunk 4]
```diff
@@ -44,28 +48,91 @@ This topic shows you how to create an API key using the Grafana UI. You can also
...
+
+#### Steps
+
+Complete the following steps to migrate from API keys to service accounts using Terraform:
+
+1. Generate `grafana_service_account` and `grafana_service_account_token` resources.
+1. When creating the service account, specify the desired scopes and expiration date.
+1. Use the token returned from `grafana_service_account_token` to authenticate the API requests.
+1. Remove references to the deprecated `grafana_api_key` resource.
+1. Track the [API keys](http://localhost:3000/org/apikeys) in use and migrate them to SATs.
```
Comment (chri2547) [inline docs/sources/administration/api-keys/index.md:138]: @eleijonmarck will the user know what this means? What are SATs?

**Review by chri2547 — CHANGES_REQUESTED** [review]
Need your full review on the migrate section, please. I did a bunch of work to refactor it. I also have some questions that I added as comments.

File: docs/sources/administration/api-keys/index.md (hunk 1 above)
Comment (eleijonmarck) [reply docs/sources/administration/api-keys/index.md:69]: il add them here

File: docs/sources/administration/api-keys/index.md (hunk 2 above)
Comment (eleijonmarck) [reply docs/sources/administration/api-keys/index.md:95]: added permissions

File: docs/sources/administration/api-keys/index.md (hunk 4 above)
Comment (eleijonmarck) [reply docs/sources/administration/api-keys/index.md:138]: removed this last section it does not make sense

File: docs/sources/administration/api-keys/index.md [hunk 5]
```diff
@@ -44,28 +48,91 @@ This topic shows you how to create an API key using the Grafana UI. You can also
...
+- xxx
+
+#### Steps
+
+Complete the following steps to migrate from API keys to service accounts using Terraform:
+
+1. Generate `grafana_service_account` and `grafana_service_account_token` resources.
+1. When creating the service account, specify the desired scopes and expiration date.
+1. Use the token returned from `grafana_service_account_token` to authenticate the API requests.
+1. Remove references to the deprecated `grafana_api_key` resource.
```
[the remark this answers is not in the document]
Comment (eleijonmarck) [reply docs/sources/administration/api-keys/index.md]: changed the wording of this

File: docs/sources/administration/api-keys/index.md [hunk 6]
```diff
@@ -27,11 +27,24 @@ Create an API key when you want to manage your computed workload with a user.
...
+To follow these instructions, you need:
+
+Either
+
+- Admin/Editor
+
+Or
+
+- API keys writer
+- Service account writer
```
Comment (eleijonmarck) [inline docs/sources/administration/api-keys/index.md:41]: @chri2547 you have a better way of presenting this? 🤔

Comment (chri2547) [reply docs/sources/administration/api-keys/index.md:41]: Is it API keys writer AND service account writer? Or just one of the two?

Comment (eleijonmarck) [reply docs/sources/administration/api-keys/index.md:41]: it's actually just service account writer/creator

File: docs/sources/administration/api-keys/index.md [hunk 7]
```diff
@@ -44,28 +57,99 @@ This topic shows you how to create an API key using the Grafana UI. You can also
...
 
-- Ensure you have permission to create Grafana service accounts. For more information about permissions, refer to [Roles and permissions]({{< relref "../roles-and-permissions/#" >}}).
+### Migrate API keys to Grafana service accounts using the Grafana user interface
 
-**To migrate all API keys to service accounts:**
+This section shows you how to migrate API keys to Grafana service accounts using the Grafana user interface. You can choose to migrate a single API key or all API keys. When you migrate all API keys, you can no longer create API keys and must use service accounts instead.
+
+#### Before you begin
+
+To follow these instructions, you need one of the following:
```
Comment (chri2547) [inline docs/sources/administration/api-keys/index.md:80]: @eleijonmarck , how does this look? Want to make sure I'm representing your feedback accurately and then we can add to all before you begin sections.

File: docs/sources/administration/api-keys/index.md [hunk 8]
```diff
@@ -44,28 +52,94 @@ This topic shows you how to create an API key using the Grafana UI. You can also
...
 1. Find the API Key you want to migrate.
 1. Click **Migrate to service account**.
+
+### Migrate API keys to Grafana service accounts using the API
+
+This section shows you how to migrate API keys to Grafana service accounts using the Grafana API.
+
+#### Before you begin
+
+To follow these instructions, you need:
```
Comment (chri2547) [inline docs/sources/administration/api-keys/index.md:105]:
```suggestion
To follow these instructions, you need one of the following:
```

File: docs/sources/administration/api-keys/index.md (hunk 3 above)
Comment (eleijonmarck) [reply docs/sources/administration/api-keys/index.md:132]: from the title of this section, it is self evident that he user needs and probably have the terraform configuration is looking for how to migrate from that. So i would say the prerequisites is not needed here.

File: docs/sources/administration/api-keys/index.md [hunk 9]
```diff
@@ -27,11 +27,19 @@ Create an API key when you want to manage your computed workload with a user.
 
 This topic shows you how to create an API key using the Grafana UI. You can also create an API key using the Grafana HTTP API. For more information about creating API keys via the API, refer to [Create API key via API]({{< relref "../../developers/http_api/create-api-tokens-for-org/#how-to-create-a-new-organization-and-an-api-token" >}}).
 
-### Before you begin:
+### Before you begin
+
+To follow these instructions, you need one of the following:
```
Comment (josmperez) [inline docs/sources/administration/api-keys/index.md:32]:
```suggestion
To follow these instructions, you need at least one of the following:
```

File: docs/sources/administration/api-keys/index.md [hunk 10]
```diff
@@ -27,11 +27,19 @@ Create an API key when you want to manage your computed workload with a user.
...
+
+- Administrator permissions
+- Editor permissions
+- Service account writer
 
-- Ensure you have permission to create and edit API keys. For more information about permissions, refer to [Roles and permissions]({{< relref "../roles-and-permissions/#" >}}).
+- To ensure you have permission to create and edit API keys. For more information about permissions, refer to [Roles and permissions]({{< relref "../roles-and-permissions/#" >}}).
```
Comment (josmperez) [inline docs/sources/administration/api-keys/index.md:38]:
```suggestion
- To ensure you have permission to create and edit API keys, follow the instructions in [Roles and permissions]({{< relref "../roles-and-permissions/#" >}}).
```

File: docs/sources/administration/api-keys/index.md [hunk 11]
```diff
@@ -45,30 +53,95 @@ This topic shows you how to create an API key using the Grafana UI. You can also
...
 1. Click **Add**.
 
-## Migrate API Keys to Grafana service accounts
+## Migrate API keys to Grafana service accounts
 
-You can migrate one or all API keys to [Grafana service accounts]({{< relref "../service-accounts/" >}}). When you migrate an API key to a service account, a service account will be created with a service account token.
-The API key will continue to work, and you can find it in the [Grafana service account tokens]({{< relref "../service-accounts/#service-account-tokens" >}}) details.
-For more information about benefits of service accounts, refer to [Grafana service account benefits]({{< relref "../service-accounts/#service-account-benefits" >}}).
+As an alternative to using API keys for authentication, you can use a service account-based authentication system. When compared to API keys, service accounts have limited scopes which provides more security than using API keys. For more information about the benefits of service accounts, refer to [Grafana service account benefits]({{< relref "../service-accounts/#service-account-benefits" >}}).
```
Comment (josmperez) [inline docs/sources/administration/api-keys/index.md:58]:
```suggestion
As an alternative to using API keys for authentication, you can use a service account-based authentication system. When compared to API keys, service accounts have limited scopes that provide more security than using API keys. 

For more information about the benefits of service accounts, refer to [Grafana service account benefits]({{< relref "../service-accounts/#service-account-benefits" >}}).
```

File: docs/sources/administration/api-keys/index.md [hunk 12]
```diff
@@ -45,30 +53,95 @@ This topic shows you how to create an API key using the Grafana UI. You can also
...
 
-You can choose to migrate a single API key or all API keys. Note that when you migrate all API keys, you can't create new API keys anymore and will have to use service accounts instead.
+The service account endpoints generate a machine-user for authentication instead of using API keys. When you migrate an API key to a service account, a service account will be created with a service account token.
```
Comment (josmperez) [inline docs/sources/administration/api-keys/index.md:60]: I think machine-user should probably be machine user. For an example in another documentation source, see https://circleci.com/docs/github-integration/#controlling-access-via-a-machine-user

File: docs/sources/administration/api-keys/index.md [hunk 13]
```diff
@@ -45,30 +53,95 @@ This topic shows you how to create an API key using the Grafana UI. You can also
...
 
-### Before you begin
+> **Note:** If you are using API keys for authentication, we recommend that you migrate your integration to the service account authentication method. The API key will continue to work. You can locate the API key in the [Grafana service account tokens]({{< relref "../service-accounts/#service-account-tokens" >}}) details.
 
-- Ensure you have permission to create Grafana service accounts. For more information about permissions, refer to [Roles and permissions]({{< relref "../roles-and-permissions/#" >}}).
+This section shows you how to migrate your integration to use the new service account endpoints. You can migrate your API keys using:
```
Comment (josmperez) [inline docs/sources/administration/api-keys/index.md:64]:
```suggestion
This section shows you how to migrate your integration to use the new service account endpoints. To migrate your API keys, use:
```

Comment (josmperez) [reply docs/sources/administration/api-keys/index.md:64]: Use all of these? One or more of these? One of these?

File: docs/sources/administration/api-keys/index.md [hunk 14]
```diff
@@ -45,30 +53,95 @@ This topic shows you how to create an API key using the Grafana UI. You can also
...
-1. Sign in to Grafana.
-1. Click **Administration** in the left-side menu and select **API Keys**.
-1. In the top of the page, find the section that says Switch from API keys to service accounts.
+### Migrate API keys to Grafana service accounts using the Grafana user interface
+
+This section shows you how to migrate API keys to Grafana service accounts using the Grafana user interface. You can choose to migrate a single API key or all API keys. When you migrate all API keys, you can no longer create API keys and must use service accounts instead.
+
+#### Before you begin
+
+To follow these instructions, you need one of the following:
```
Comment (josmperez) [inline docs/sources/administration/api-keys/index.md:76]: one or at least one?

File: docs/sources/administration/api-keys/index.md [hunk 15]
```diff
@@ -45,30 +53,95 @@ This topic shows you how to create an API key using the Grafana UI. You can also
...
+- Editor permissions
+- Service account writer
+
+For more information about permissions, refer to [Roles and permissions]({{< relref "../roles-and-permissions/#" >}}).
+
+#### Steps
+
+To migrate all API keys to service accounts, complete the following steps:
+
+1. Sign in to Grafana, hover your cursor over **Configuration** (the gear icon), and click **API Keys**.
```
Comment (josmperez) [inline docs/sources/administration/api-keys/index.md:88]:
```suggestion
1. Sign in to Grafana, point to **Configuration** (the gear icon), and click **API Keys**.
```

File: docs/sources/administration/api-keys/index.md [hunk 16]
```diff
@@ -45,30 +53,95 @@ This topic shows you how to create an API key using the Grafana UI. You can also
...
 1. Find the API Key you want to migrate.
 1. Click **Migrate to service account**.
+
+### Migrate API keys to Grafana service accounts using the API
+
+This section shows you how to migrate API keys to Grafana service accounts using the Grafana API.
+
+#### Before you begin
+
+To follow these instructions, you need one of the following:
```
Comment (josmperez) [inline docs/sources/administration/api-keys/index.md:107]: one or at least one?

File: docs/sources/administration/api-keys/index.md [hunk 17]
```diff
@@ -45,30 +53,95 @@ This topic shows you how to create an API key using the Grafana UI. You can also
...
+
+### Migrate API keys to Grafana service accounts using Terraform
+
+This section shows you how to migrate API keys to Grafana service accounts using Terraform.
+
+#### Before you begin
+
+To follow these instructions, you need:
+
+- N/A
```
Comment (josmperez) [inline docs/sources/administration/api-keys/index.md:138]: Is this truly Not Applicable? Or is it a placeholder for info to be added later?

**Review by josmperez — DISMISSED** [review]
Suggestions for Grafana style.

File: docs/sources/administration/api-keys/index.md (hunk 9 above)
Comment (eleijonmarck) [reply docs/sources/administration/api-keys/index.md:32]: it is true that you only need at least one

File: docs/sources/administration/api-keys/index.md (hunk 17 above)
Comment (eleijonmarck) [reply docs/sources/administration/api-keys/index.md:140]: as long as you have access to your terraform configuration, anyone that has that (regardless of the permissions in Grafana)

File: docs/sources/administration/api-keys/index.md (hunk 13 above)
Comment (eleijonmarck) [reply docs/sources/administration/api-keys/index.md:66]: use all of these steps. i think maybe you are confused by the word use here? 

if we say "To migrate your API keys:" would this make more sense?

File: docs/sources/administration/api-keys/index.md (hunk 12 above)
Comment (eleijonmarck) [reply docs/sources/administration/api-keys/index.md:62]: changed to machine user

File: docs/sources/administration/api-keys/index.md (hunk 13 above)
Comment (eleijonmarck) [reply docs/sources/administration/api-keys/index.md:66]: this is really dependant on the customer. if they have API calls using API keys, or terraform using API keys they need to migrate through that section, does this make sense?

File: docs/sources/administration/api-keys/index.md [hunk 18]
```diff
@@ -27,11 +27,19 @@ Create an API key when you want to manage your computed workload with a user.
...
+- Editor permissions
+- Service account writer
 
-- Ensure you have permission to create and edit API keys. For more information about permissions, refer to [Roles and permissions]({{< relref "../roles-and-permissions/#" >}}).
+- To ensure you have permission to create and edit API keys, follow the instructions in [Roles and permissions]({{< relref "../roles-and-permissions/#" >}}).
 
-**To create an API key:**
+### Steps
+
+To create an API
```
Comment (chri2547) [inline docs/sources/administration/api-keys/index.md:42]:
```suggestion
To create an API, complete the following steps:
```

File: docs/sources/administration/api-keys/index.md [hunk 19]
```diff
@@ -45,30 +53,103 @@ This topic shows you how to create an API key using the Grafana UI. You can also
...
+For more information about the benefits of service accounts, refer to [Grafana service account benefits]({{< relref "../service-accounts/#service-account-benefits" >}}).
 
-### Before you begin
+The service account endpoints generate a machine user for authentication instead of using API keys. When you migrate an API key to a service account, a service account will be created with a service account token.
 
-- Ensure you have permission to create Grafana service accounts. For more information about permissions, refer to [Roles and permissions]({{< relref "../roles-and-permissions/#" >}}).
+> **Note:** If you are using API keys for authentication, we recommend that you migrate your integration to the service account authentication method. The API key will continue to work. You can locate the API key in the [Grafana service account tokens]({{< relref "../service-accounts/#service-account-tokens" >}}) details.
 
-**To migrate all API keys to service accounts:**
+## This section shows you how to migrate Grafana to using service accounts.
```
Comment (chri2547) [inline docs/sources/administration/api-keys/index.md:66]:
```suggestion
## Ways of migrating API keys to service accounts
```

File: docs/sources/administration/api-keys/index.md (hunk 3 above)
Comment (chri2547) [reply docs/sources/administration/api-keys/index.md:142]: Gotcha. Removing.

File: docs/sources/administration/api-keys/index.md [hunk 20]
```diff
@@ -45,30 +53,103 @@ This topic shows you how to create an API key using the Grafana UI. You can also
...
+
+1. SATs used for authentication
+1. Remove code that handles the old `/api/auth/keys` endpoint.
+1. Track the [API keys](http://localhost:3000/org/apikeys) in use and migrate them to SATs.
+
+### Migrate API keys to Grafana service accounts using Terraform
+
+This section shows you how to migrate API keys to Grafana service accounts using Terraform.
+
+#### Before you begin
```
Comment (chri2547) [inline docs/sources/administration/api-keys/index.md:142]:
```suggestion
```

File: docs/sources/administration/api-keys/index.md [hunk 21]
```diff
@@ -45,30 +53,103 @@ This topic shows you how to create an API key using the Grafana UI. You can also
...
+
+To follow these instructions, you need at least one of the following:
+
+- N/A
```
Comment (chri2547) [inline docs/sources/administration/api-keys/index.md:146]:
```suggestion
```

**Review by chri2547 — COMMENTED** [review]
Added suggestions.

**Review by chri2547 — APPROVED** [review]

**Pull request merged**